pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/migrating/xml/ForLoopCanBeForeach.xml

```diff
@@ -225,4 +226,42 @@
         ]]></code>
     </test-code>
 
+    <test-code>
+        <description>Consider iterators only if safe</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+        class Foo {
+          void loop() {
+            Iterable<DataFlowNode> path = null;
+
+            for (Iterator<DataFlowNode> i = path.iterator(); i.hasNext();) {
+              DataFlowNode inode = i.next();
+              path.remove(inode); // throws ConcurrentModificationException if it were a foreach
+              if (inode.getVariableAccess() == null) {
+                continue;
+              }
+            }
+          }
+        }
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Do not report iterator loop if we can't find iterator decl</description>
```

Comment:
I'm not sure this is to be expected... sounds more as a limitation to the current implementation... am I missing something?

Reply:
But the current typeresolution code *can* actually resolve the type of inherited fields (no need to get their declaration!)... so we can actually know if `path` is `Iterable` provided a complete classpath.

The only reason for not being able to detect it would be an incomplete classpath (missing classes), which would yield `null` as the type for path.

If we depended less on symbol table and more on type resolution directly, more of these scenarios would be addressed.

Reply:
Hum well the expression `path.iterator()` is parsed as a primary prefix with image `path.iterator` and a primary suffix for the arguments. I think we cannot get the type of `path` there, only that of `path.iterator()`, right?

Reply:
You are right, we need #497 for this... it will have to wait then.